source/common/network/connection_impl.cc

```diff
@@ -421,7 +421,11 @@ void ConnectionImpl::onReadReady() {
   IoResult result = transport_socket_->doRead(read_buffer_);
   uint64_t new_buffer_size = read_buffer_.length();
   updateReadBufferStats(result.bytes_processed_, new_buffer_size);
-  onRead(new_buffer_size);
+  if (result.bytes_processed_ != 0) {
```

Comment:
This seems reasonable to me. I can't think of any reason why we would want to raise onRead if no bytes were processed, even if there are bytes still in the read buffer. However I want to make sure I understand the flow here which I don't think I do. Can you clarify the sequence of events that causes the problems on OSX that don't happen on Linux? I'm mainly trying to understand if there is potentially a larger lurking problem here that we don't know about. @alyssawilk can you take a look at this also?

Reply:
I'm curious as well.  Is this only on initial connection (like the initial EPOLLIN arrives before bytes and somehow translates to onRead?) or in general?

Reply:
The sequence is:

1. Read event. In this test 36 bytes (`"HTTP/1.1 200 OK\r\ncontent-length: 0\r\n"`) are added to `read_buffer_` and `Http::MockClientConnection::dispatch()` is invoked (via `onRead`). The test does not arrange to drain the buffer.
2. A second read event is triggered. There's no new data, but `onRead` is invoked again and the same buffer is delivered to dispatch. I believe in actual usage the http parser would have processed the bytes and the buffer would have been drained.
3. `Http::CodecClient` invokes dispatch with an empty buffer as part of handling the stream reset.

Step 2 doesn't always happen. I think it's related to the socket close. AFAICT libevent should be using kqueue on OSX. Based on my reading of the kqueue man page, kqueue read events are triggered on remote close, which sets an EV_EOF flag in the kqueue event. This may happen with or without available data. Looking at libevent they don't seem to check for this case. I presume that usually there's one read event with data and EV_EOF set, but sometimes the data and the EOF are reported separately and libevent passes both events on.

My other branch modifies this test to accept that the buffer might be delivered once or twice. It's also not difficult to modify the test to drain the buffer, but we'd still see 3 calls to dispatch (one with data, and two with empty buffers).

Reply:
Ahhh, so basically on MacOS we were just getting the FIN separately and the current code would do an onRead for that?  Makes sense and code LGTM.

Could I also ask that you dup the flaky test where we did a write + close and event loop and instead write (loop) close (loop) to unit test how this is likely to occur in production?

Reply:
I think this looks good to me also. In the case where we process zero bytes, can we also assert that `result.action_ == PostIoAction::Close`? I think that should always be the case. I think in general this is a better more common solution, as we don't require connection users to drain read buffers.

Reply:
I don't trust either linux or macos to never cause a spurious wakeup, given that they're mostly harmless.  I think asserting is too harsh.

Reply:
@mattklein123 in SSL connections a handshake error (see `source/common/ssl/ssl_socket.cc` lines 38 and 103) results in PostIoAction::KeepOpen but no data. Looking at adding the assertion in RawBufferSocket, but there's already a case in there that handles EAGAIN, so it may well be possible to get an empty read on an open socket there too.

Reply:
OK just skip it.

Reply:
And empty read with KeepOpen also happens when the read buffer has reached it's buffer limit (if I'm reading the test correctly).